python/chemiscope/input.py

```diff
@@ -156,32 +92,182 @@ def write_input(path, frames, meta=None, extra=None, cutoff=None):
+
+    This function uses :py:func:`create_input` to generate the input data, see
+    the documentation of this function for more information.
+    """
+
+    if not (path.endswith(".json") or path.endswith(".json.gz")):
+        raise Exception("path should end with .json or .json.gz")
+
+    data = create_input(frames, meta, properties, cutoff)
+
+    if "name" not in data["meta"] or data["meta"]["name"] == "<unknown>":
+        data["meta"]["name"] = os.path.basename(path).split(".")[0]
+
     if path.endswith(".gz"):
         with gzip.open(path, "w", 9) as file:
             file.write(json.dumps(data).encode("utf8"))
     else:
         with open(path, "w") as file:
             json.dump(data, file)
+
+
+def _typetransform(data, name):
+    assert isinstance(data, list) and len(data) > 0
+    if isinstance(data[0], str):
+        return list(map(str, data))
+    elif isinstance(data[0], bytes):
+        return list(map(lambda u: u.decode("utf8"), data))
+    else:
+        try:
+            return [float(value) for value in data]
+        except Exception:
+            raise Exception(
+                f"unsupported type in property '{name}' values: "
+                + "should be string or number"
+            )
+
+
+def _linearize(name, property, n_structures, n_atoms):
+    """
+    Transform 2D arrays in multiple 1D arrays, converting types to fit json as
+    needed.
+    """
+    data = {}
+    if isinstance(property["values"], list):
+        data[name] = {
+            "target": property["target"],
+            "values": _typetransform(property["values"], name),
+        }
+    elif isinstance(property["values"], np.ndarray):
+        if len(property["values"].shape) == 1:
+            data[name] = {
+                "target": property["target"],
+                "values": _typetransform(list(property["values"]), name),
+            }
+        elif len(property["values"].shape) == 2:
+            if property["values"].shape[1] == 1:
+                data[name] = {
+                    "target": property["target"],
+                    "values": _typetransform(list(property["values"]), name),
+                }
+            else:
+                for i in range(property["values"].shape[1]):
+                    data[f"{name}[{i + 1}]"] = {
+                        "target": property["target"],
+                        "values": _typetransform(list(property["values"][:, i]), name),
+                    }
+        else:
+            raise Exception("unsupported ndarray property")
+    else:
+        raise Exception(
+            f"unknown type ({type(property['values'])}) for property '{name}'"
+        )
+
+    if "units" in property:
+        for item in data.values():
+            item["units"] = str(property["units"])
+
+    if "description" in property:
+        for i, item in enumerate(data.values()):
+            # add [component XX] to the description if values was a ndarray
```

Comment:
why?

Reply:
To transform something like `PCA of SOAP vector` to `PCA of SOAP vector [component 1]` when associated with the `PCA[1]` property. I don't have strong feeling on this, we can remove it.